src/Microsoft.Diagnostics.NETCore.Client/HandleableCollection.cs

```diff
@@ -0,0 +1,316 @@
+        /// Returns the first item offered to the handler
+        /// or waits for a future item if no item is immediately available.
+        /// </summary>
+        /// <param name="token">The token to monitor for cancellation requests.</param>
+        /// <returns>A task that completes with the first item offered to the handler.</returns>
+        public Task<T> HandleAsync(CancellationToken token)
+        {
+            VerifyNotDisposed();
+
+            return HandleAsync(DefaultHandler, token);
+        }
+
+        /// <summary>
+        /// Returns the item on which the handler completes and waits for future items
+        /// if the handler does not immediately complete.
+        /// </summary>
+        /// <param name="handler">The handler that determines on which item to complete.</param>
+        /// <param name="token">The token to monitor for cancellation requests.</param>
+        /// <returns>A task that completes with the item on which the handler completes.</returns>
+        public Task<T> HandleAsync(Handler handler, CancellationToken token)
+        {
+            VerifyNotDisposed();
+
+            return HandleAsync(
+                handler,
+                tcs => tcs.TrySetCanceled(token),
+                token);
+        }
+
+        private async Task<T> HandleAsync(Handler handler, Action<TaskCompletionSource<T>> tokenCallback, CancellationToken token)
+        {
+            var completionSource = new TaskCompletionSource<T>(TaskCreationOptions.RunContinuationsAsynchronously);
+            using var methodRegistration = token.Register(() => tokenCallback(completionSource));
+            using var disposalRegistration = _disposalSource.Token.Register(
+                () => completionSource.TrySetException(new ObjectDisposedException(nameof(HandleableCollection<T>))));
+
+            lock (_items)
+            {
+                OnHandlerBegin();
+
+                bool stopHandling = false;
+                for (int i = 0; !stopHandling && i < _items.Count; i++)
+                {
+                    T item = _items[i];
+
+                    stopHandling = TryHandler(item, handler, completionSource, out bool removeItem);
+
+                    if (removeItem)
+                    {
+                        _items.RemoveAt(i);
+                        i--;
+                    }
+                }
+
+                if (!stopHandling)
+                {
+                    _handlers.Add(Tuple.Create(completionSource, handler));
+                }
+            }
+
+            return await completionSource.Task.ConfigureAwait(false);
+        }
+
+        private static bool TryHandler(in T item, Handler handler, TaskCompletionSource<T> completionSource, out bool removeItem)
+        {
+            removeItem = false;
+            if (completionSource.Task.IsCompleted)
+            {
+                return true;
+            }
+
+            bool stopHandling = handler(item, out removeItem);
```

Comment:
What kind of exception safety (if any) should we have here? Right now an exception in any handler will break the whole loop

Reply:
I changed it to catch the exception and allow the handler to still be registered.